src/overlays/actors/ovl_Boss_03/z_boss_03.c

```diff
@@ -41,157 +266,2321 @@ const ActorInit Boss_03_InitVars = {
+            Math_ApproachS(&this->morphTable[i].x, (temp_f14 - temp_f0) * 200.0f, 0xA, phi_s0);
+            if (phi_s0 != 0) {
+                phi_s3 |= 1;
+            }
+        }
+    }
+
+    if (phi_s3 & 1) {
+        Math_ApproachS(&this->actor.shape.rot.y, Math_FAtan2F(zDiff, xDiff), 0x14, 0x800);
+    }
+
+    if (sGyorgInstance->actor.world.pos.y - 40.0f < sGyorgInstance->waterHeight) {
+        for (i = 0; i < 6; i++) {
+            xDiff = sGyorgInstance->actor.world.pos.x - this->unk_2DC[i].x;
+            yDiff = sGyorgInstance->actor.world.pos.y - this->unk_2DC[i].y;
+            zDiff = sGyorgInstance->actor.world.pos.z - this->unk_2DC[i].z;
+
+            temp_f0 = sqrtf(SQ(xDiff) + SQ(yDiff) + SQ(zDiff));
+
+            if ((i == 0) && (temp_f0 > 400.0f)) {
+                break;
+            }
+
+            phi_s0 = sGyorgInstance->actor.speedXZ * 16.0f;
+            temp_f14 = sGyorgInstance->actor.speedXZ * 5.0f + 150.0f;
+            if (phi_s0 > 0x1000) {
+                phi_s0 = 0x1000;
+            } else if (phi_s0 < 0x100) {
+                phi_s0 = 0x0100;
+            }
+
+            if (temp_f0 < temp_f14) {
+                Math_ApproachS(&this->morphTable[i].x, (temp_f14 - temp_f0) * 200.0f, 0xA, phi_s0);
+                if (phi_s0 != 0) {
+                    phi_s3 |= 2;
+                }
+            }
+        }
+
+        if (phi_s3 & 2) {
+            Math_ApproachS(&this->actor.shape.rot.y, Math_FAtan2F(zDiff, xDiff), 0x14, 0x800);
+        }
+    }
+
+    if (phi_s3 == 0) {
+        for (i = 0; i < 6; i++) {
+            Math_ApproachS(&this->morphTable[i].x, 0, 0x14, 0x80);
+        }
+    }
+}
+
+Gfx* D_809E91C0[] = {
+    gGyorgSeaweedTopDL,    gGyorgSeaweedPiece5DL, gGyorgSeaweedPiece4DL,
+    gGyorgSeaweedPiece3DL, gGyorgSeaweedPiece2DL, gGyorgSeaweedPiece1DL,
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Boss_03/func_809E6B70.s")
+void Boss03_SeaweedDraw(Actor* thisx, GlobalContext* globalCtx) {
+    Boss03* this = THIS;
+    s16 i;
+    // Why 10 Mtxs? This seems to only use the first 6 elements
+    Mtx* mtx = GRAPH_ALLOC(globalCtx->state.gfxCtx, sizeof(Mtx) * 10);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Boss_03/func_809E6BC0.s")
+    OPEN_DISPS(globalCtx->state.gfxCtx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Boss_03/func_809E6CB4.s")
+    func_8012C28C(globalCtx->state.gfxCtx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Boss_03/Boss03_Update.s")
+    gSPSegment(POLY_OPA_DISP++, 0x0D, mtx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Boss_03/func_809E7920.s")
+    Matrix_InsertTranslation(this->actor.world.pos.x, this->actor.world.pos.y, this->actor.world.pos.z, MTXMODE_NEW);
+    Matrix_Scale(this->actor.scale.x, this->actor.scale.y, this->actor.scale.z, MTXMODE_APPLY);
+    Matrix_RotateY(this->actor.shape.rot.y, MTXMODE_APPLY);
+    Matrix_InsertZRotation_s(0x4000, MTXMODE_APPLY);
+    // The indices looks a bit random...
```

Comment:
I assume that's part of the point: it's meant to make it naturally wavy, right?

Reply:
:man_shrugging: